filesys/handler.go

```diff
@@ -0,0 +1,300 @@
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+	switch r.URL.Path {
+	case "/fs/stat":
+		fa.handle(&Stat{}, w, r)
+	case "/fs/fstat":
+		fa.handle(&Fstat{}, w, r)
+	case "/fs/open":
+		fa.handle(&Open{}, w, r)
+	case "/fs/write":
+		fa.handle(&Write{}, w, r)
+	case "/fs/close":
+		fa.handle(&Close{}, w, r)
+	case "/fs/rename":
+		fa.handle(&Rename{}, w, r)
+	case "/fs/readdir":
+		fa.handle(&Readdir{}, w, r)
+	case "/fs/lstat":
+		fa.handle(&Lstat{}, w, r)
+	case "/fs/read":
+		fa.handle(&Read{}, w, r)
+	case "/fs/mkdir":
+		fa.handle(&Mkdir{}, w, r)
+	case "/fs/unlink":
+		fa.handle(&Unlink{}, w, r)
+	case "/fs/rmdir":
+		fa.handle(&Rmdir{}, w, r)
+	default:
+		fa.doError("not implemented", "ENOSYS", w)
+	}
+}
+
+type Responder interface {
+	WriteResponse(fa *Handler, w http.ResponseWriter)
+}
+
+func (fa *Handler) handle(responder Responder, w http.ResponseWriter, r *http.Request) {
+	if err := json.NewDecoder(r.Body).Decode(responder); err != nil {
+		fa.logger.Printf("ERROR handle : %s\n", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if fa.debug {
+		fa.logger.Printf("handle %s %+v\n", r.URL.Path, responder)
+	}
+	responder.WriteResponse(fa, w)
+}
+
+type ErrorCode struct {
+	Error string `json:"error"`
+	Code  string `json:"code"`
+}
+
+func (fa *Handler) doError(msg, code string, w http.ResponseWriter) {
+	if fa.debug {
+		fa.logger.Printf("doError %s : %s\n", msg, code)
+	}
+	e := &ErrorCode{Error: msg, Code: code}
+
+	w.WriteHeader(http.StatusBadRequest)
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(e); err != nil {
+		fa.logger.Println("doError json error :", err)
+	}
+}
+
+func (fa *Handler) okResponse(data any, w http.ResponseWriter) {
+	if marshal, err := json.Marshal(data); err != nil {
+		fa.logger.Println("okResponse json error:", err)
+		w.WriteHeader(http.StatusInternalServerError)
+	} else {
+		if fa.debug {
+			fa.logger.Printf("okResponse %s\n", string(marshal))
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		_, _ = w.Write(marshal)
+	}
+}
+
+func fixPath(path string) string {
+	return strings.TrimPrefix(path, "/fs/")
+}
+
+type Stat struct {
+	Path string `json:"path,omitempty"`
+}
+
+type Open struct {
+	Path  string `json:"path"`
+	Flags int    `json:"flags"`
+	Mode  uint32 `json:"mode"`
+}
+
+func (o *Open) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	fd, err := syscall.Open(fixPath(o.Path), o.Flags, o.Mode)
+	if fa.handleError(w, err, true) {
+		return
+	}
+	response := map[string]any{"fd": fd}
+	fa.okResponse(response, w)
+}
+
+type Fstat struct {
+	Fd int `json:"fd"`
+}
+
+type Write struct {
+	Fd       int    `json:"fd"`
+	Buffer   string `json:"buffer"`
+	Offset   int    `json:"offset"`
+	Length   int    `json:"length"`
+	Position *int   `json:"position,omitempty"`
+}
+
+func (wr *Write) WriteResponse(fa *Handler, w http.ResponseWriter) {
+
+	if wr.Position != nil || wr.Offset != 0 {
```

Comment:
Do we plan to implement those in a future PR? I do see a `syscall.Seek` for the Read endpoint for position.

Reply:
This was mainly a carry over from the original wasm_exec.js write

https://github.com/golang/go/blob/451e4727ec825a7ce6f6e6f82761ff90c33fec83/misc/wasm/wasm_exec.js#L27-L34

The read endpoint seek was implemented due to the coverage tool performing reads with offsets. 
i.e. coverage metadata generation was broke without it. 

I'll look and see how feasible it is to implement it here also.